### ISSUE TITLE
Out of curiosity - Performance

### ISSUE THREAD

**Issue by jonathan-s** [EmilStenstrom/django-components#14]
There is a similar library for rails -> https://viewcomponent.org/ (https://github.com/github/view_component) Apparently that library is 10x faster in rendering than partials, which is the equivalent of django's include. 

Do you have any idea what the performance characteristics are for this library?

**Comment by EmilStenstrom** [comment]
@jonathan-s That is a great question. I haven't don't any benchmarking, but I would expect similar performance metrics as using a normal include. This would be easy to test I think. If you have a template that includes something in a loop, maybe several steps deep, and then at the bottom level, you put an include and a {% component %} call, and compare the loading time in a structured way.

The motivation for this library was not performance, but rather developer ergonomics in working with components. That said, I would happily accept patches that 1) adds a benchmark and 2) improved performance.

**Comment by EmilStenstrom** [comment]
I think we should add a benchmark test in the form of a python script that renders a complex template 1000 times, first with include and then with component, and compares the results. If anyone is able to contribute such a script I'd gladly merge it.

**Comment by mlissner** [comment]
Performance is one of the biggest concerns I have while evaluating this. My site has a number of pages with hundreds of links that we intend to convert to components. Right now, the links are generated in a Django template loop. 

If we move to components, and each link requires hitting disk to load the component, we're in a lot of trouble. I'm going to go dig into the documentation on this, but I thought I'd highlight how this is an important priority (at least for me).

Thank you!

**Comment by EmilStenstrom** [comment]
@mlissner happy to include a benchmark in the library if you write one!

**Comment by mlissner** [comment]
Fair! We'll look into it.

**Comment by mlissner** [comment]
@EmilStenstrom What do you have in mind for the benchmark? Are you thinking of just some documentation here in the issue, a Github Action that runs on each PR, or what? I'm thinking for our purposes, we could just do a comparison of includes vs. components, and document it here. 

I think it's also worth mentioning performance on the main documentation page and linking to the caching docs. Performance was one of my biggest concerns.

**Comment by mlissner** [comment]
We've done a Github Action for one of our repos that you can see here: 

https://github.com/freelawproject/reporters-db/pull/190#issuecomment-2518773885

It's nice b/c it runs on every PR and so in theory you can see the before and after of the PR. It's a lot of work to maintain though (Github is constantly changing permissions, etc). 

---

We'll wait to see if Emil has any comments, and then we'll see if we can do a little script to output performance. I think the important part will be publishing the results though. Maybe the documentation should have something on the Welcome page that says it's performant, and then links to the caching page, which could have the results (assuming they're good!)?

**Comment by EmilStenstrom** [comment]
It sounds like a couple of different (and independently valuable) steps:
1. Get a rough idea of performance running a script manually
2. Publishing results to build trust
3. Cementing performance levels with GitHub action 

Starting with 1 sounds good to me!

**Comment by JuroOravec** [comment]
Mentioning this here as we definitely need to check also the parsers once it comes to performance analysis - https://github.com/EmilStenstrom/django-components/pull/908#discussion_r1917361641

**Comment by JuroOravec** [comment]
Btw as part of my job I was looking at the performance of our Django app. The slowest page has about 167kb, uses 13-17 different components, and renders in total 363 component instances. And it takes ~8.0s to load.

Turns out that rendering components is ~7.5 out of those 8s.

That's.... that's pretty bad 😄 

I did some profiling with Silk, and if I'm reading it correctly, then 100% and 95% of the time is spent in [`postprocess_component_html`](https://github.com/EmilStenstrom/django-components/blob/7ed4fd88f981a94e8b008a394b65138fa404ca44/src/django_components/dependencies.py#L330) and  [`_link_dependencies_with_component_html`](https://github.com/EmilStenstrom/django-components/blob/7ed4fd88f981a94e8b008a394b65138fa404ca44/src/django_components/dependencies.py#L273) respectively.

<img width="742" alt="Image" src="https://github.com/user-attachments/assets/072c3ac2-f816-457f-a398-ece83ab93d9b" />

# Root cause

Those functions are used to insert component IDs as HTML attrributes, so we can do scoped CSS, JS / CSS vars, and similar features. Currently we use BeautifulSoup4 for that.

However, the problem is that when components are inserted one into another, then the HTML content is re-parsed each time it's nested.

E.g. If I had 3 components, A, B, C, where C was nested in B and B nested in A:

```txt
- CompA
   |- CompB
        |- CompC
```

Then the process is following:

1. CompA starts rendering, pauses to render its children
2. CompB starts rendering, pauses to render its children
3. CompC starts rendering, pauses to render its children
4. Content of CompC is rendered.
5. We pass content of CompC to BS4, to apply `data-djc-id-a1b2df` (and similar)
6. CompC is done
7. Content of CompB is rendered. Content of CompB already includes the rendered content of CompC
8. We pass content of CompB to BS4, to apply `data-djc-id-a1b2df` (and similar) - as part of this, CompC is re-parsed the 2nd time
9. CompB is done
10. Content of CompA is rendered. Content of CompA already includes the rendered content of CompB and CompC
11. We pass content of CompA to BS4, to apply `data-djc-id-a1b2df` (and similar) - as part of this, CompB is re-parsed 2nd time, while CompC is re-parsed 3nd time.

So as you can see, the more deeply the components are nested, the more the duplicate parsing there occurs.

As a bit of a mental exercise, if we assume that:
1. All components are roughly of the same size
2. That the parsing speed is constant across the whole template (no shortcuts)
3. That each component contains 2 other components
    
Then what we get is that, if there's 3 levels of depth, then we already spend 80% of the time parsing duplicate HTML. In other words, we could get away with rendering only 20% of the content if we changed the approach:

```
1                                                                            - 0/1 repeats
1  2a  2a  2b  2b                                                            - 2/5 repeats
1  2a  2a  2b  2b 3aa 3aa 3aa                                                - 10/17 repeats
                   3ab 3ab 3ab
                   3ba 3ba 3ba
                   3bb 3bb 3bb
1  2a  2a  2b  2b 3aa 3aa 3aa   4aaa 4aaa 4aaa 4aaa                          - 34/43 repeats
                                4aab 4aab 4aab 4aab
                   3ab 3ab 3ab  4aba 4aba 4aba 4aba
                                4abb 4abb 4abb 4abb
                   3ba 3ba 3ba  4baa 4baa 4baa 4baa
                                4bab 4bab 4bab 4bab
                   3bb 3bb 3bb  4bba 4bba 4bba 4bba
                                4bbb 4bbb 4bbb 4bbb
```

I traced my case, where the component structure looks roughly like this:

```
- Page
  |- Layout 
    |- Layout2
      |- Header
        |- Breadcrumbs
      |- Navbar
        |- Icon
        |- UserLogin
      |- Sidebar
        |- Bookmarks
     (main content)
      |- List
      |- Tabs
        |- TabItem (5x)
          |- Component with biz logic
            |- Expansion panel
              |- Biz logic (N times based on number of db entries)
                |- Other biz logic (4x)
```

Where the very last entry "Other biz logic (4x)" is nested in 8 components. So IMO it's safe to say that there's a lot of time wasted parsing the HTML that was already parsed.

# How to fix this

## Design-wise

### ❌ Idea 1 - Defer serialization

Problem is that the interface for passing around the rendered data is plain string. In other words, when I render a component with `Component.render`, I can't predict if the result still nested in another Component, or if the result is already meant to be rendered in a vanilla Django template.

I think that if we didn't allow components to be used in vanilla Django templates, then we would be able to pass the rendered data as its own kind of object, instead of a string. Think of this special object as similar to Vue's [VNode](https://vuejs.org/guide/extras/render-function#creating-vnodes). And if we passed around this object instead of a string, we could defer serializing it. That way we could get rid of the issue that the HTML of the nested components are serialized and re-serialized again and again.

However, I don't think this is currently feasible with django-components. Because it would practically mean that one would not be able to use `{% component %}` inside a vanilla Django template. Instead, `{% component %}` would be allowed only inside other component templates. And top-level would have to be rendered as `Component.render` / `Component.render_to_response`. (Kinda similar pattern to dealing with async functions).

### ✅ Idea 2 - Render components top-down instead of bottom-up (???)

Maybe we could render change the rendering flow, so that we apply `data-djc-id-...` BEFORE we render the child components.

E.g. one approach could be to do similar to what I've seen in django-cotton, where they mask some parts of the HTML, and then parse / transform it, and then un-mask the resulting HTML.

So if I had a component template like this:

```django
<div>
  <span>
    {% component "table" / %}
  </span>
</div>
```

Then we'd render it, but store the output somewhere else, and return only a placeholder (similar to how the `{% component_js_dependencies %}` returns a placeholder:

```django
<div>
  <span>
    <component name="__COMPONENT__123__" />
  </span>
</div>
```

And only then apply the HTML attributes:

```django
<div data-djc-id-a1b2c3>
  <span>
    <component name="__COMPONENT__123__" />
  </span>
</div>
```

And afterwards, we'd insert the rendered content back

```django
<div data-djc-id-a1b2c3>
  <span>
    <!-- _RENDERED "table_309dcf,31c0da" -->
    <table>
      <tbody>
        ...
      </tbody>
    </table>
  </span>
</div>
```

Since the placeholder would be an HTML tag, this could work even with passing the `data-djc-id-a1b2c3` down the components. E.g. if the component was defined at the root:

```django
<div>
</div>
{% component "table" / %}
```

Then the placeholder would receive the `data-djc-id-a1b2c3` too:

```django
<div data-djc-id-a1b2c3>
</div>
<component name="__COMPONENT__123__" data-djc-id-a1b2c3 />
```

And thus we'd know that we should apply `data-djc-id-a1b2c3` to all root elements of the nested component (in ADDITION to their own `data-djc-id-`:

```django
<div data-djc-id-a1b2c3>
</div>
<!-- _RENDERED "table_309dcf,31c0da" -->
<table data-djc-id-31c0da data-djc-id-a1b2c3>
  <tbody>
    ...
  </tbody>
</table>
```

## Implementation-wise

Currently we use BeautifulSoup4 for parsing the HTML. However, in theory, for the sake of what django-components needs HTML parsing for, there could be an implementation that assigns `data-djc-id-a1b2c3` and similar in a single pass through the HTML text. This is described in https://github.com/EmilStenstrom/django-components/issues/831.

I don't expect the custom implementation to be necessarily faster..

Buuut... Once there would be such own implementation, I'd be possible to move it to Rust with PyO3 for potential 10x improvement (I don't expect crazy optimizations, but even a naive translation could improve things, [see this blog post](https://ohadravid.github.io/posts/2023-03-rusty-python/)).

I assume that the custom parser And django-components could 

---

All in all, the combined approaches could theoretically give a 50x improvement. So the page I tested it on would render only in 0.65s instead of 8s.

@EmilStenstrom What are your thoughts?

**Comment by mlissner** [comment]
Thank you so much for this analysis. It's really helpful to learn this since we're starting our implementation now. If rendering a fairly simple page like you describe takes that long, it's probably a show stopper for us. Even the 50× improvement still hurts, really. We try to render in under one second and some of our pages contain hundreds of items ([here's an example](https://www.courtlistener.com/docket/4214664/national-veterans-legal-services-program-v-united-states/)). If the SSR part of our pipeline takes more than a fraction of a second, that's not great.

Sorry if this message is a bit demoralizing. Let's find a way to make this fast. We'd rather not get deep into this library since our expertise lies elsewhere, but we might be able to if that's the best way to make this happen. We do have some engineers that are pretty good at performance profiling.

That said, are do y'all have suggestions for how we can help? And if so, how quickly do we think this could be done, since it feels like a pretty big overhaul?

Thank you again! We're excited to make this library great, if we can.

**Comment by JuroOravec** [comment]
Hey @mlissner thanks for response!

Let's take it point by point:

### Fixing slow rendering

I too have responsibilities towards my employer, so this is a high priority for me. This and next week I was intending to work on polishing stuff around templating, so that can be pushed to later. I'm in middle of one feature, which I plan to finish tomorrow or on Saturday, then I'll be on this.

I expect that the design change would take 2-3 days (assuming it's feasible). This would be those 5x improvement.

Next would be replacing BeautifulSoup with custom parser - Here too I assume it might take 2-3 days. I already have half-working parser. However, if BeautifulSoup's HTML parser is highly optimized, this could again worsen the performance a bit.

However, as mentioned in previous post, having the custom parser unlocks the possibility to move it to Rust for the performance gains. I guess that could take me a week - although I would know the Python parser code, I'm novice with Rust, and haven't used PyO3 / maturin yet.

So overall I expect about 2 weeks of work.

### Performance requirements

> Even the 50× improvement still hurts, really. We try to render in under one second and some of our pages contain hundreds of items ([here's an example](https://www.courtlistener.com/docket/4214664/national-veterans-legal-services-program-v-united-states/)). If the SSR part of our pipeline takes more than a fraction of a second, that's not great.

My contributions to this library were with the intention to enable fast development, performance was an afterthought (if a thought at all). There's still some features that I'd like to get in that might impact performance. So the performance will likely change (worsen), and there will need to be more rounds of optimisations.

The point I'm getting to is that @mlissner it sounds like you have a good idea of what performance is or is not acceptable to you.

We also can't improve what we don't measure.

And so at this point it'd be useful to just build out the test suite to be able to measure performance. If you have some pages that are a good representation of your usage, we could add those as tests. I think ideally tests would include two variants
- One with vanilla django templates (I assume you are using those right now?)
- The other with the page built with django-components

> Let's find a way to make this fast. We'd rather not get deep into this library since our expertise lies elsewhere, but we might be able to if that's the best way to make this happen. We do have some engineers that are pretty good at performance profiling.
>
> That said, are do y'all have suggestions for how we can help? And if so, how quickly do we think this could be done, since it feels like a pretty big overhaul?

If you've got some engineers that could help, that'd be amazing!

- If you or somebody from your org could help us set up the benchmarking tests, that'd help a lot. Especially if that person would set it up in such a way that it would be meaningful for you.

- After that, if there's someone on the team with experience in Rust (or similar), they could take on the conversion of the HTML parser script from Python to Rust. I expect that someone knowledgeable in the language could make it substantially more performant than me. I assume that could take up to a week - to understand the original script, to rewrite it, to test it and to publish it.

After these two steps, I think it would make sense to then re-evalue what are your performance needs, and if django-components is not meeting them, then how it could be achieved. And I think only at this point it would maybe make sense to do profiling of the codebase.

@EmilStenstrom What are your thoughts?

**Comment by mlissner** [comment]
That's great news. I'm checking if we have any Rust experts (Crustaceans, I think they're called?). I don't think we have anybody on our team, but we have about 100 devs in our slack, so it's possible (but unlikely). 

We can help develop a small benchmark. I imagine you'd want it ASAP, right, so you can do the other work?

**Comment by JuroOravec** [comment]
Great to hear! And nope, the benchmark and the Rust port aren't blocking me :) But of course, the sooner the better!

**Comment by EmilStenstrom** [comment]
@JuroOravec Great that you found this early, super helpful to have that kind of benchmark to try with.

About the HTML parser: I'm a little worried about the maintenance of having another programming language and dealing with binary packages as part of django-components. So some thoughts:
* Are we really sure we should do this with a HTML parser? I know we need this in the future, but could we fall back to regexes and revisit in our own pace later?
* If we are going the html route, my experience from writing my own html parser, is that the majority of the time ends up in error handling, fixing mismatched tags and other atrocities that are allowed in HTML. If we require valid HTML, and break on errors, that parser will be MUCH faster. Likely fast enough in python.

@mlissner Thanks for being understanding. This is a pre-1.0 project, so things like these are bound to happen unfortunately. Juro is the expert here, and is working on consulting basis I think, so maybe hire him a couple of days to focus on this? No matter what, a benchmark where we can compare vanilla Django and django-components from a performance standpoint would be great.

**Comment by mlissner** [comment]
Have y'all looked at lxml? It's supposed to be very fast as well. (I also have concerns about using Rust, but if it's just a pip away, I can live with that, probably.)

**Comment by JuroOravec** [comment]
> Are we really sure we should do this with a HTML parser? I know we need this in the future, but could we fall back to regexes and revisit in our own pace later?

@EmilStenstrom I wanted to hear your opinion first before putting in too many questions or thoughts, but yes, maybe a Python regex-based implementation will be sufficient.

For a Python implementation, I think it could live within django-components. But if we chose to go with Rust, then I guess it would make more sense as a standalone package / repo.

---

@mlissner I did try lxml. Also [selectolax](https://github.com/rushter/selectolax). Issue with both is that they modify the HTML. As Emil said, they try to fix it. Or, at least in case of selectolax, that parser is built to mimick the browser parser / the official HTML spec, and that introduces unwanted behaviour like wrapping HTML fragments in `<html><head></head><body></body></html>`.

And that's problematic, because individual components don't have to output a valid HTML. Only once the output of them all is combined, only then it should become valid HTML.

**Comment by mlissner** [comment]
Right, got it, sorry, I remember seeing that in a different issue. Thanks for the reminder.

**Comment by JuroOravec** [comment]
Update:
1. I got the design change implemented. Looks like, in the end, this saved only about 20%. The page that was previously loading at 8s now still loads at 8s on first load, and ~6.5s on second load. Though I'm testing these outside of Docker, and I noticed that outside of Docker the loading times are a lot more variable.
   - On first load, 60% is spend on BeautfiuSoup parsing, 20% is spent in our `tag_parser.py` - so also the latter one should be refactored too. When I wrote `tag_parser.py` I used a loop that goes character by character, because I wasn't aware that using regexes is much more efficient in Python. But for now I will focus on the HTML parser.
   - On second load, results of `tag_parser.py` are already cached, so 80% are spent on BeautfulSoup parsing.

2. I had a go at the custom Python HTML parser. The naive implementation I had stashed around resulted in render time of 60s. But with LLMs I was able to get it down to 6.2s on first render and 5s on second, but out of that the HTML parser is still 4s. So it's slightly better than BeautifulSoup.

3. Other things to try:
   - Currently I run the HTML parser for each rendered component, and then once more at the very end with the full HTML. I think the very last call could be skipped.
   - There's maybe 2-3 ideas I want to still try with the Python parser implementaton.
   - If those ideas won't lead to a considerable improvement, then I'll explore the going low-level with Rust.

**Comment by EmilStenstrom** [comment]
@JuroOravec Awesome that you are doing progress on this. Remind me again, why do we need a html parser? I'm thinking that there might be a workaround that we can switch to temporarily, until we have time to optimize perf.

**Comment by JuroOravec** [comment]
It's used for adding HTML attributes like
- `data-djc-id-...` - Needed for passing JS variables
- `data-djc-css-...` - Needed for passing CSS variables
- `data-djc-scope-...` - Needed for CSS scoping

**Comment by EmilStenstrom** [comment]
And more specifically to find the outermost (plural) elements in a template to add the ID to?

**Comment by JuroOravec** [comment]
## HTML attributes

> And more specifically to find the outermost (plural) elements in a template to add the ID to?

@EmilStenstrom Yes, so:
1. `data-djc-id-` would be applied to all the root elements of the component's HTML template. The ID is component **instance** specific.
2. `data-djc-css-` same location as `data-djc-id-`, but the CSS id is component **class** specific. So all components of the same class would have the same `data-djc-css-` ID.
3. `data-djc-scope-` this would be applied to all component's OWN HTML elements - So we first fin all root elements (like in case of `data-djc-id-`, and then walk down the HTML tree, until we come across a nested component.

   However, fill slots are also scoped, if defined from within the template. A good way how it's described for Vue is that you scope those HTML elements that you "can see defined within HTML template"
   ```html
   <div>  <!-- apply scope -->
     <span>  <!-- apply scope -->
       {% component "button" / %}  <!-- stop scope -->
         <span>  <!-- apply scope -->
       {% endcomponent %}
     </span>
   </div>
   ```

---

## Rust implementation

So I had a go with Rust, and I was able to reduce the html parsing time to ~0.1s for the demo page of 170 kB. (~70-80x improvement overall).

<img width="285" alt="Image" src="https://github.com/user-attachments/assets/95527c50-648d-411f-9409-1b227cc4aa8c" />

<img width="285" alt="Image" src="https://github.com/user-attachments/assets/fa3412b7-98bf-4863-84f4-bebb67ddedb3"/>

Notes:
- This is already compiled for production.

- These are python tests using the python bindings of the Rust code.

- These tests are different than the profiling I did a few days back. So I have yet to see how the performance will look within Django.

- I still have to read up on how to publish [maturin / PyO3](https://pyo3.rs/v0.18.1/#using-rust-from-python) packages.

- I'm sure that the Rust implementation could be further optimized by Rust experts. But I think, for a non-expert like me, this is as good as it gets right now.

---

## Allow to disable JS / CSS features for extra speed?

As for @mlissner's case, assuming we integrate the Rust implementation, then @mlissner, if you need to trim down even those 0.1s, I'm thinking we could add a setting to globally disable the JS / CSS features (or at least CSS scoping), in which case we wouldn't need to add any of the HTML attributes like `data-djc-id-`.

- Technical detail: `data-djc-scope-` and `data-djc-css-` are not problematic, they can be omitted if not need. But `data-djc-id-` must be always present if *any* of the ancestor components may use CSS scoping . Because `data-djc-id-` is also used so that we know when to stop applying `data-djc-scope-`. 

---

## Next steps

### 1. HTML parser

Personally I'm for using the Rust implementation.

Cons:
- It's much less likely that newcomers / passer-bys will be able to help maintain it, unlike the Python code.

Pros:
- The performance is just too good to pass up on.
- It's IMO fairly stable - I don't see the need to change the Rust implementation anytime soon - none of the features that I want to implement down the line will need any changes.
- And if it actually comes to making changes, it's actually fairly doable to do adjustments when one uses LLM assistants (which IMO looks like such assistants will become part of the software engineer's toolset, the same way as we might now look with horror at someone who doesn't use IDEs).

But IMO it should live in a separate package, so that django-components would just use it as a dependency.

The django-components will include tests for the API of this package, so we double check that any future swaps or modifications will not interfere with django-components.

### 2. Move django-components to an org?

If the HTML parser is a separate package, IMO it makes sense for it to live close to django-components. While the parser could be used also standalone, it's primarily meant to fix issues for django-components. So it feels weird if the repo lived under e.g. `JuroOravec/djc-html-parser`.

So I think it might be relevant to revisit the idea about moving django-components to an organization, so that the HTML parser would still be visibly managed by django-components.

This would allow also other future such optimisations to live under django-components umbrella.

### 3. Optimizing other parsers

#### Tag parser

As mentioned earlier in this thread, currently in my demo page that was loading 8s, 1s of that was spent in `tag_parser.py`.
This tag parser takes a template tag within Django template, e.g. `{% component ... %}`, and allows to parse syntax like `attr:key=val`, literal lists and dicts (`key=[1, 2, 3]`), spread operators (`...var`), etc.

To this parser, I only plan to add support for comments `{# ... #}`, but don't see a need for more.

Since this is again the case of parsing string, I'd also move it to Rust, so that the performance becomes a non-issue.

#### CSS parser

Next, for the scoped CSS feature, in the initial implementation we might use regex to prefix CSS selectors, so e.g.
  
  `div.my-class > p` 
  
   becomes

   `div.my-class[data-djc-css-a3bc36] > p[data-djc-css-a3bc36]`

But this might trip up on comments or media queries like [`@media`](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_media_queries/Using_media_queries).

So if regex will be insufficient, I would try out a CSS parser  like [tinyCSS](https://github.com/Kozea/tinycss2).

But this might end up being the same case as with BeautifulSoup, where the the python implementation is simply too slow. And in such case I can imagine moving the string transformation to Rust.


### 4. Other optimizations?

When I was looking at the profiling results for my 170 kB page with 360 components, it's possible that around 1s of the time was spent in non-parser code.

Thing are blurry, because there were also some database requests being made from within the components. 

But, in the hypothetical case that the Python side of rendering components takes up to 1s, that's still not ideal. But also, this is the code that's under active development. So I'd leave further optimisations for later, once the code has stabilised.

**Comment by mlissner** [comment]
You are making fast progress, sorry we haven't been able to keep up. I don't have much to add, but I'll just say that the idea of disabling JS/CSS for added performance sounds pretty OK to me. If I understand correctly, this would disable *inline* CSS and JS, right? If so, that's something we don't do anyway (I think we even block it with CORS), so it'd be entirely fine by me, and even welcome to block that as a policy level. (Occasionally OSS volunteers submit inline CSS, and we have to reject their code; it'd be great if their code never even worked!)

We'll be using tailwind, so we'll have a zillion classes on our components, but I assume those would still work.

Thank you again for all the hard work, this is so cool to see coming together!

**Comment by JuroOravec** [comment]
@mlissner Sorry forgot to explain. But no, the inlined JS / CSS would still work.

What I had in mind are the not-yet-implemented features shown here https://github.com/EmilStenstrom/django-components/issues/836 - So it's
1. Being able to pass Python variables to JS or CSS,
2. Being able to run JS function (with the data from Python) when a component is rendered on the page.
3. Vue-like CSS scoping.

**Comment by mlissner** [comment]
Ok, that's neat. We won't need that for 99% off pages, and we can work around it for the last 1%, so I think turning it off would be our preference if it gave more performance. 

The best for us might be the option to default it off, but enable it on certain components, but that probably easier to say than do!

Thank you!

**Comment by EmilStenstrom** [comment]
@JuroOravec "Allow to disable JS / CSS features for extra speed?" - Could this be dependent on if the get_js_data and get_css_data methods are there? As for scoped or not, we could store that in the registry for fast access.

About Rust: I'm still worried about the maintenance burden of this. Could you share the benchmark file to see if it's possible to hack something together with regex and skip HTML parsing instead?

**Comment by JuroOravec** [comment]
## RE: Disabling JS / CSS features for extra speed

> "Allow to disable JS / CSS features for extra speed?" - Could this be dependent on if the get_js_data and get_css_data methods are there? As for scoped or not, we could store that in the registry for fast access.

Yup, there is already couple of things set up such that they are included / generated only if the JS / CSS data is there. To be precise, the method is always there (or may be defined on parent class), but I check if those methods returned data dictionary, and whether that dictionary is non-empty. If it's empty, there's no variables, and we don't generate anything.

~But we can exclude only the `data-djc-css-` this way when there is no CSS data. As I said earlier, to be able to exclude `data-djc-id-`, there would have to be no JS data, AND disabled CSS scoping.~

...

Silly me. Actually with the changes I did over the weekend, things are a bit different.

One of the optimizations was to ensure that we pass component's rendered HTML to the HTML parser only once.

How I did it was that, during render, nested components output a placeholder instead of their actual content.

So this

```django
<div>
  <h1>Hi</h1>
  {% component "table" / %}
</div>
```

gets converted to this

```html
<div>
  <h1>Hi</h1>
  <template djc-render-id="abf63d"></template>
</div>
```

This way we can walk down the tree and apply the scoping attribute:

```html
<div data-djc-scope-a17f62>
  <h1 data-djc-scope-a17f62>Hi</h1>
  <template djc-render-id="abf63d" data-djc-scope-a17f62></template>
</div>
```

It doesn't affect child components, because child components are not part of this snippet. As a counter example, previously, the child's HTML was first rendered, and then placed directly in place of `{% component .. %}`

So this

```django
<div>
  <h1>Hi</h1>
  {% component "table" / %}
</div>
```

Became

```html
<div>
  <h1>Hi</h1>
  <table data-djc-id-ab26f3>
    <tbody>
      <tr>
        <td>123</td>
      </tr>
    </tbody>
  </table>
</div>
```

In the old implemenetation, I needed the `data-djc-id-ab26f3` to know that I must NOT apply `data-djc-scope-a17f62` to the `<table>` and its descendants.


### Implications

- 👍 So that means that we don't need `data-djc-id-` to control where the CSS scoping is being applied.
- 👍 So we don't need to disable CSS scoping

- ❓ However, I realized there's one more usage of `data-djc-id-`, and that's for the `$onLoad` JS callback (see https://github.com/EmilStenstrom/django-components/issues/836). There we will need the HTML attribute to find the corresponding HTML element in the DOM. One of my intentions with it is that this would then allow a clean integration with AlpineJS components.

  - 👀 But maybe what we could do is to check the Component.js for the `$onLoad` keyword. And if it IS there, then we MUST set `data-djc-id-`.
 
---

So overall, the logic would be:
- `data-djc-css-` is applied only if `get_css_data()` returns non-empty dict
- `data-djc-id-` is set if EITHER:
  - `get_js_data()` returns non-empty dict
  - `$onLoad` is found within Component.js
- `data-djc-scope-` is applied only when `Component.css_scope` is `True`

Thus, if none is true, then we wouldn't have to apply any special attributes, and we could skip the HTML parsing.


## Rust impl

> About Rust: I'm still worried about the maintenance burden of this. Could you share the benchmark file to see if it's possible to hack something together with regex and skip HTML parsing instead?

First of all, sad news, my Rust parser is actually shite :'(

Turns out that the HTML page that I'm rendering at work, it has a total size of 2.2 MB. It was 170 kB compressed, but the server actually generates 2.2 MB of HTML.

When I crank my benchmarks up to 2 MB:
- Python
  - Read only - 28s
  - Read and write - 48s
- Rust
  - Read only - 13s
  - Read and write - 80s (lol)

Now, in all cases this is much more than the 6-8s I have to wait to load my page. So even though it's 2MB, then, when loading an actual page, that HTML probably has much fewer HTML elements, since that's when I called a callback function, and that's probably the slow part.

Also, partially what probably slows Rust down is that kept the implmenetation one-to-one, meaning that on each `<elem>` in the HTML, Rust code called up a Python function.

So instead, next step I'd try to:
1. Do away with the Python callback.
2. Let [quick-xml](https://crates.io/crates/quick-xml) handle HTML reading / writing.
3. I'd thus write only logic that allows to insert arbitrary HTML attributes to 1) root elements, and 2) all elements.

Based on the parsing speed listed [here](https://github.com/RazrFalcon/roxmltree#parsing) (2,997,887 ns/iter per 800kb), I'm hoping the overall perf will be in the order of milliseconds.

And lastly, as for the benchmark, here's a single Python file that contains both benchmarking logic and fastest Python implementation I got:

[benchmark_py.txt](https://github.com/user-attachments/files/18482501/benchmark_py.txt)

But IMO it's still not simple nor reliable with regex. The problem sure got simpler now that we don't have to check each node for `data-djc-id-` to know whether to keep applying the same HTML attribute to the descendants or not. But we still need to be able to target only root elements. HTML allows comments, and string quotes and JS syntax, and all behave bit differently, so one needs to track which context they are in. And at that point it IMO quickly gets complex fast, and you end up with the same file as I got 😄 But happy to be proven wrong.

**Comment by JuroOravec** [comment]
We're so back! 0.1s to process 2MB of HTML, inserting the HTML attributes along the way. And only on 150 lines of code.

<img width="373" alt="Image" src="https://github.com/user-attachments/assets/8bb6c5bb-ab98-4c06-aa73-2647f89325e3" />

<img width="200" alt="Image" src="https://github.com/user-attachments/assets/485d46bd-b35a-4044-875b-8f3826221276" />

**Comment by EmilStenstrom** [comment]
@JuroOravec Awesome! So we’re switching from bs4 till quick-xml and that’s it?

**Comment by JuroOravec** [comment]
The previous results weren't the optimized production build. With the production build, we're at 0.063s per 5MB.

<img width="282" alt="Image" src="https://github.com/user-attachments/assets/54e0cb32-1c18-45a5-ae95-339dbbf680d0" />

---

> @JuroOravec Awesome! So we’re switching from bs4 till quick-xml and that’s it?

Yup, that's the idea, but to clarify:

### 1. Build and publish the HTML parser

quick-xml and the implementation are still written in Rust, so it still requires building and publishing that. Overall the Rust project looks like this (ignore the files in `tests`)

<img width="282" alt="Image" src="https://github.com/user-attachments/assets/dccb0281-f59c-4b33-a10f-182a72008f87" />

- `CI.yml` - Build & publish Github action auto-generated by maturin. it creates wheel for linux, musllinux  (Alpine Linux), Windows, and MacOS
- `html_parser.rs` -  Actual implementation + Rust tests
- `lib.rs` - Package entrypoint, imports `html_parser.rs`
- Then `__init__.pyi`, `pyproject.toml` and `py.typed` are for the generated Python package.
- And `Cargo.toml` and `Cargo.lock` are for Rust project + dependencies

Looks like the project template is quite straightforward. We just need to agree on where to put this (what repo and package name),  and the github actions should take care of the rest.

### 2. Refactor django-components

As mentioned before, I made some changes to django-components like replacing `{% component ... %}` with temporary placeholders.

As a first step, I want to push those changes. And as part of that, also replace BS4 with the best performing custom Python HTML parser.

The second step would then be to replace the Python parser with this Rust implementation.

The reason for using the custom Python parser as the middle-step is because both that custom Python parser and Rust implementation use a streaming approach*. Whereas BS4 first builds a whole tree / DOM, and then you make changes upon it.

\* AKA you give a callback or otherwise specify what happens at specific HTML tokens, and that's how you make modifications.

**Comment by dalito** [comment]
Not sure what it requires or how much complication it brings, but could you maybe keep a slower pure python version and make using the optimized rust-library opt-in?

**Comment by EmilStenstrom** [comment]
Oh, I thought it was just a python release of and existing quick-xml package?

**Comment by EmilStenstrom** [comment]
@JuroOravec I've moved the repo to an organization, and invited you as an owner of the organization. The URL has now changed.

**Comment by JuroOravec** [comment]
Sorry, I got side-tracked with a proof of concept for moving the tag_parser.py to Rust too, to shred off the 1 second on the first load. There I chose a more declarative approach, using [pest](https://docs.rs/pest/latest/pest/#pest-the-elegant-parser) package that allows to write a parser by declaring the tokens in a [BNF](https://en.wikipedia.org/wiki/Backus%E2%80%93Naur_form)-like syntax. There I still need to fix literal dictionaries before it's fully working.

@EmilStenstrom Just saw that, thanks for inviting! Looks like the links to the deployed docs broke ([this works](https://django-components.github.io/django-components/latest/overview/welcome/), but [this doesn't](https://emilstenstrom.github.io/django-components/latest/)). In the evening I can go over the links in the codebase / docs to fix that.

After that, today or tomorrow, I'll get back to adding the new repo for the HTML parser.

---

## Check potential recursion limitations

Btw, in terms of performance / reliability, I realized there might be one more limitation for large projects, in terms of recursion. 

I'd have to check, but when one renders a component from within the template tag using `{% component ... %}`, then a single call like this may generate 6-7 levels of calls. E.g. it might look like:

 `template.render()-> template._render() -> NodeList.render() -> ComponentNode.render() -> Component._render() -> Component._render_impl()`

Normally python allows recursion up to 1000 levels, which would allow components of depth of ~140 layers. So this should be fine for 99% of the projects. But I can imagine that that if there was a component like a [treeview](https://dribbble.com/tags/treeview), which would generate nested component depending on given data, then that limit could be reached.

So I want to check if this could be an issue.

With the changes I made with regards to optimistions (that I yet have to make a PR for), I think we could make further changes so that, instead of rendering components recursively, we'd have a render stack, rendering componenents depth-first and first-in-first-out. I think this would allow for practically infinite depth of components (limited only by the device's memory).

I made a ticket for that here https://github.com/django-components/django-components/issues/923

**Comment by JuroOravec** [comment]
One more idea: 

TL;DR: The refactor to support infinitely nested components (#923) would also pave the way for us to be able to aggressively cache rendering on component level. Other optimizations I mentioned until now, like the Rust implementation and whatnot, they are reducing the cost of the EXTRA features that django-components offers. So they just make it sure that django-components is comparable in speed with vanilla Django templates, or other packages. BUT, this caching could actually make websites using django-components substantially faster (on subsequent loads).

Building on previous ideas, I plan to refactor the component rendering, so that this:

```py
class MyComp(Component):
    template = """
      <div>
        {% component "table" / %}
        <span> Hello </span>
        {% component "button" %}
           Click me!
        {% endcomponent %}
      </div>
    """
```

Would be actually rendered into this, with a placeholder where the components were:

 ```py
class MyComp(Component):
    template = """
      <div>
        <template djc-render-id="a1b23c"></template>
        <span> Hello </span>
        <template djc-render-id="f61d33"></template>
      </div>
    """
```

Now, we could go a step further and do such placeholder also for slots. So I the original template contained a slot:

```py
class MyComp(Component):
    template = """
      <div>
        {% component "table" / %}
        <span>
          {% slot "content" default / %}
        </span>
        {% component "button" %}
           Click me!
        {% endcomponent %}
      </div>
    """
```

Then we could put a placeholder for the slot too, to get:

 ```py
class MyComp(Component):
    template = """
      <div>
        <template djc-render-id="a1b23c"></template>
        <span>
          <slot name="content" djc-render-id="c3c3c3"></slot>
        </span>
        <template djc-render-id="f61d33"></template>
      </div>
    """
```

Next, we'd need a user to give us a promise that the component's rendered output depends ONLY on the input - the args and kwargs (potentially also context).

For starts, it could be opt-in, e.g.:

```py
class MyComp(Component):
    cache = True
```

Later (v1), this could become the default, so that users get caching out of the box.

Now, at this point we could establish a direct cause-and-effect - given inputs produce given output. So we could hash the inputs, and store the output.

Thanks to #923, we'd already have the code in place such that nested components are not embedded into the HTML of the parent prematurely, but all is put together at the end. That means that, at the point of rendering and putting together the components, we could first hash the inputs, and check if the component with given input is already cached. Whether we used a cached result or not, we'd then take the rendered template and search and replace the component placeholders (`<template djc-render-id="a1b23c">`), as well as slot placeholders (`<slot name="content" djc-render-id="c3c3c3">`), to arrive at the final output.

Since this would be done on a component basis, then I think for building blocks UI components like buttons, icons, and whatnot, the cache would be highly used. Probably the same for layout components, as those also don't tend to change much. It would only be the "business logic" pages and components that would be better of uncached.

Here we could potentially go a step further, such that if user explicitly set `cache = True/False`, we'd respect that. But if not, or they set it to `cache = None`, then we could have some heuristics that would decide whether to cache given component or not, e.g. based on the ratio of hits and misses on the first 50-100 calls.

The complex part of this would be handling of slots.
- Slot fills that were defined within the template with `{% fill %}` tag could be treated as static text, IF the parent component where the `{% fill %}` was used cached and reused too.
- If the slot fill accessed slot's data, this would be more complex, because we'd need to know the state of both the parent and child components (and whether they support caching), to be able to reliably say that the fill content can be reused. Otherwise, these this slot fill would be handled as a function, and rendered on each call.
- As for slot fill that were passed in from Python via `Component.render()`, these would be always rendered on each call (tho there could be some option for users to opt into caching)
- Potentially, for slots that opt-in, there could be similar logic - that if the inputs to the slot function haven't changed, then we could reuse the previously-rendered slot fill.

---

Another potential approach to caching could be to figure out which variables are being used in te template

**NOTE: After I finished writing this section, I realized that this is NOT ❌ a good idea for caching. But I keep the notes below in case we ever need to extract knowledge of used variables inside a template.**

1. For each `{% tag ... %}`, we'd check the inputs that are passed to it.
   1. Unfortunately, I think we'd be able to do so only with our `{% component %}` tag, and possibly with a small set of Django's own tags ([IF](https://docs.djangoproject.com/en/5.1/ref/templates/builtins/#if), [FOR](https://docs.djangoproject.com/en/5.1/ref/templates/builtins/#cycle), [WITH](https://docs.djangoproject.com/en/5.1/ref/templates/builtins/#with)), because Django's tags IMO don't have a (sufficiently) standardized inputs. For example, the [`{% lorem 3 w %}`](https://docs.djangoproject.com/en/5.1/ref/templates/builtins/#lorem) tag accepts `w` to set the mode, but if I passed plain `w` to [`{% firstof var1 var2 w %}`](https://docs.djangoproject.com/en/5.1/ref/templates/builtins/#firstof) it would be understood as a variable named `w`. 
   2. So if we came across a `{% ... %}` tag that is NOT supported, we'd NOT cache the component, because the tag could be [`{% now %}`](https://docs.djangoproject.com/en/5.1/ref/templates/builtins/#now).
2. For each variable call like `{{ my_var|default:default_var }}`, we'd check for the variables used. As part of django-components, I've already written a tag parser, and that could be used to figure out which variables are used. Here I think it's sensible to assume that filters should be predictable and without side effects ([pure functions](https://en.wikipedia.org/wiki/Pure_function)).
3. Don't know how I feel about `{% include %}` and `{% extend %}`, but my hunch is to say that the component would NOT be cached if we came across these. So basically we'd approach these as the UNSUPPORTED `{% ... %}` tags.

**Comment by JuroOravec** [comment]
## Lil status update:

TL;DR - After I finish the template tag parser, I expect the loading time to be 0.9 - 1.1s for a 2MB of HTML (360 components) on my machine.

- ✅ The Rust-based HTML parser was implemented and integrated.

- ✅ The component rendering logic was redesigned so that each templates is rendered and post-processed only ever once.

- 🚧 I'm finishing working on Rust-based template tag parser to shave off 1s from first load.

- ❌ I will keep the logic that applyies HTML attributes like `data-djc-id-a1b2c3`

   We apply HTML attributes like `data-djc-id-a1b2c3` to the roots of rendered HTML. [Earlier in this thread](https://github.com/django-components/django-components/issues/14#issuecomment-2603245172), I was discussing skipping this if there's no need for it:
   > So overall, the logic would be:
   > 
   > - data-djc-css- is applied only if get_css_data() returns non-empty dict
   > - data-djc-id- is set if EITHER:
   >   - get_js_data() returns non-empty dict
   >   - $onLoad is found within Component.js
   > - data-djc-scope- is applied only when Component.css_scope is True
   > 
   > Thus, if none is true, then we wouldn't have to apply any special attributes, and we could skip the HTML parsing. 

   However IMO it doesn't make sense anymore to remove that, because:
   - After migrating the HTML parser to Rust, this now takes only 1-2% of the whole render time.
   - It is useful for debugging to have those HTML attributes, as it shows component boundaries. 
   - Removing it means updating about half of all the tests, which is currently not worth it right now for those 1-2%.

## Is there more to do?

### Perf leads

   I looked through the profiler data (generated with cProfiler via [django-silk](https://github.com/jazzband/django-silk), then exported and visualized in [snakeviz](https://jiffyclub.github.io/snakeviz/))

   <details>
   <summary> (Click to expand) Some functions that caught my eye </summary>

   <img width="252" alt="Image" src="https://github.com/user-attachments/assets/9b0737e2-b3f3-42c6-bbeb-33afc93ae53e" />
    <img width="280" alt="Image" src="https://github.com/user-attachments/assets/e49fe4c4-989f-4228-9dd5-60b6f1e4ab3c" />
    <img width="281" alt="Image" src="https://github.com/user-attachments/assets/0bdce9d3-7d40-4fae-95c3-71ba312d7d34" />
    <img width="278" alt="Image" src="https://github.com/user-attachments/assets/5b022da4-0f5e-438b-b081-4a36f4714f7f" />
    <img width="270" alt="Image" src="https://github.com/user-attachments/assets/6fdd48f4-cd93-4596-89b2-9f4ef2647d25" />
    <img width="256" alt="Image" src="https://github.com/user-attachments/assets/b742cfbd-6e17-4efa-baef-436e721891f4" 
    />
    <img width="281" alt="Image" src="https://github.com/user-attachments/assets/8bc09fc7-8122-4a80-bc95-da14c4d8fb74" />

   </details>

   However, most of what remains is stuff that takes 2.5-3% of render time. So across them it could be maybe 10-15% - IF they can even be optimized at all.

   Moreover, of those 0.9 - 1.1s:
   - 0.1s was spent calling `{built-in method posix.stat}`
   - ~0.1s was spent accessing data from `Context` object

   So we're getting into a zone where further speed gains take more effort and more nuance.

### Component-level caching

@EmilStenstrom I think it makes sense to keep the idea of caching components open. But maybe not in the way I described before, but simply as an alterantive to Django's `{% cache %}`. Because for example in @oliverhaas's example, he wrapped a component in `{% cache %}` tag:

```django
{% cache 3600 foo %}
  {% component "icon" name="check" / %}
{% endcache %}
```

And I think it should be possible to set this on the component class itself (though API's not clear yet).

### My outlook

   Personally I think that it makes sense to again focus on features. After django-components matures a bit more again, then there might pop up some new opportunities. E.g. the `component_post_render` in `perfutil/component.py` is fairly well isolated. So in the future we could potentially refactor out the main render loop and move it to lower level.

 I also think that there will be optimization opportunities if we take over parsing of whole templates.
   - In such case we could for example pre-compile parts of the template to avoid needless resolution.  So if there's a template like this:
     ```django
     {% for item in menu_list %}
       {{ item }}
     {% endfor %}
     {% component "test" / %}
     ```

     And Component looks like this:
     ```py
     menu_list = ["a", "b", "c", ...]

     class Table(Component):
         def get_context_data(self):
             return {
                 "menu_list": menu_list,
             }          
     ```

     Then with a bit of Python-fu, it WOULD be possible to detect that `menu_list` is constant (does not change with component inputs). And with this info, we could modify the Django template, inlining the static data:
     ```django
       a
       b
       c
     {% component "test" / %}
     ```
     Thus avoiding having to render the `{% for %}` loop again and again.

## Establish benchmark

After finishing the template tag parser, I think it makes sense to focus on this:

### 1. Standardized django-components example

So all my performance data are just relative. What 8s or 0.9s to me might not be to you. Also, the page I test the perf on has a bug where I access the database from within the template. So the 0.9s-1.1s is just an estimate (actual time is 1.4s, and I assume database interactions take 0.3-0.5s).

So @mlissner this comes back to our previous discussion on having some tests that would represent real-life usage.

I'm also thinking about converting my test page into such test. Though I wonder what would be a good way to write that. In my case there's I think 13 or 17 different components. Theoretically I could put them all in a single file, inlining the HTML, and then have a function that creates synthetic data, and renders it all. It would be a huge test file, but at least it would be self-contained.

@EmilStenstrom What do you think?

### 2. Standardized django example

Moreover, I have no idea how django-components compares against vanilla Django or other options. After having the django-components example in place, it would be good to convert it to vanilla django, and see how fast that renders.

**Comment by mlissner** [comment]
Sorry I (and the devs on my team) haven't been any help here yet. We've gotten bogged down on a client job. Bah.

I think the important thing with benchmarks is that they represent reality at least a bit. I think big test cases are fine for that. We can contribute ours when we get a little further along too, but we don't have much to share yet. 

The question I have is what does it mean to fail that test case though? How do you track the average test case across PRs vs. the code in a given PR to decide that it has failed? 

What we do (which I shared earlier) is run a performance test against the PR and `main`. It makes a comment [like this](https://github.com/freelawproject/eyecite/pull/200#issuecomment-2631434540): 

![Image](https://github.com/user-attachments/assets/71d3b6d6-2655-4cfd-9076-36836268b60a)

If you look closely, there are two lines there in the chart that are right on top of each other — blue is `main`, yellow is the PR. The PR didn't degrade or enhance performance, so you can just barely see the blue. Sometimes though, the lines diverge meaningfully. 

So, I don't have ideas on tracking performance across a series of PRs, but this helps highlight major changes at least?

**Comment by JuroOravec** [comment]
> The question I have is what does it mean to fail that test case though? How do you track the average test case across PRs vs. the code in a given PR to decide that it has failed?

Yup, that's the big question.

I searched a bit, and came across [airspeed-velocity (asv)](https://github.com/airspeed-velocity/asv), used by numpy and such. I would need to look more into it, but the main selling point to me is that:
1. It does benchmarking
2. It can track results across releases (or across commits?)
3. It generates static HTML (I think.. There's an option to [view results in the browser](https://pv.github.io/numpy-bench/#/)) - So we could embed it into the docs site.
4. It's open source, no need for any 3rd party service.

To contrast, there's also:
- a pytest plugin [pytest-benchmark](https://pytest-benchmark.readthedocs.io/en/stable/usage.html), but that one doesn't generate HTML.
- SaaS products like [Codspeed](https://codspeed.io/home#product), which might have some publically accessible UI (didn't check). But who knows if they will be around in a couple of years, whereas local tool will be indefinitely.

I so think it might be worth a shot to look into avs. Figure out how to set it up, and check if it could be deployed as part of docs site.

Firstly it would be just about making the performance data visible, to give people some sort of ballpark of what to expect. In [the demo](https://pv.github.io/numpy-bench/#/) you can see that there's loads of tests. I think for starts it might be sufficient have just 4:
1. Small example (1-3 components) implemented in vanilla Django templates
2. Small example implemented in django-components
3. Large example (e.g. example with 360 components) in vanilla Django
4. Large example in django-components

**Comment by JuroOravec** [comment]
Btw, on another note, In https://github.com/django-components/django-components/issues/14#issuecomment-2631557434 I mentioned it might be worthwhile to check the time spent retrieving data from the Context object.

I looked into that in https://github.com/django-components/django-components/pull/955. It turned out to be red herring:
 - my implementation added a whole 1 second to the render, so 10x slower than original.
- I also wasn't able to find my previous claim about "~0.1s was spent accessing data from Context object".

I created and closed the PR to document the code and share how it can be done, but I won't pursue this further.

**Comment by mlissner** [comment]
I feel like putting visibility into the documentation site isn't going to be very helpful. You'll be able to compare this vs. django templates, but what we really want is something in the PR itself that says, "Hey, this degraded performance!" 

To that end, I like the pytest thing, but of course, it will break tests when your Github Runner is slow, and that's unfortunate. Because of that, you'll want to loosely constrain your test (give a 1s-long test 1.5s, say) so it doesn't trigger all the time, at that point your test isn't really doing much.

I don't have any better answers, aside from the thing I shared before, but I admit it's a lot to get it working.

**Comment by EmilStenstrom** [comment]
I think it makes sense with performance test that fail if the performance degrades by a certain percent compared to main. 

Important that the test measures many different kinds of performance, with a focus on what you'll likely see in the wild. So I'd say let's set up a diverse set of tests with different kinds of component usage, and start tracking that.

WHEN we have a test set we like, we can integrate it into the PR workflow. I don't think it will be that difficult. 

pytest-benchmark looks good I think. With `--benchmark-histogram` it seems to generate an SVG that can be embedding in a PR. I think that's likely the simplest way to get this going in a nice way?

**Comment by JuroOravec** [comment]
> I feel like putting visibility into the documentation site isn't going to be very helpful. You'll be able to compare this vs. django templates, but what we really want is something in the PR itself that says, "Hey, this degraded performance!"

> I think it makes sense with performance test that fail if the performance degrades by a certain percent compared to main.

> pytest-benchmark looks good I think. With --benchmark-histogram it seems to generate an SVG that can be embedded in a PR. I think that's likely the simplest way to get this going in a nice way?

@mlissner @EmilStenstrom  I think it's two different needs. Testing perf on PRs is a good idea.

But also, as a maintainer, I want to see the long-term trends. E.g. if we allowed 5% perf deterioration per PR, that could still degrade the package gradually over the long term. Also correct me if wrong, but looking only at the PyPI page, it looks like pytest-benchmarks measures only time taken, not memory usage. I also want to know how django-components fares memory-wise against vanilla Django, or if I accidentally nuke memory if I introduce a new feature. Hence why I'd like to have also something more comprehensive like the `asv`.

> Could we use code search to find some more examples of tricky cases? https://github.com/search?type=code&q=%22%7B%25+component%22+language%3APython+

Feel free to search and add suggestions.

I'm trying to make a test case out of my work project (anonymized), as my case is IMO pushing the boundaries of django-components a lot. I just got it working for the "isolated" mode. But there's again some kinda bug with how the slots work with the "django" mode, which is really frustrating. Then I'd try to make a copy of it that'd use `{% include %}` and `{% extends %}` to measure the vanilla Django case.

What's strange is that the in-test render time is about 3x faster than in live project (250ms in test vs 750ms live when comparing second page load). But if this same strange behaviour will be present also when testing the vanilla Django case, then it's fine to compare the two.

**Comment by JuroOravec** [comment]
Also, might be better as a separate ticket later on, but it'd be useful to see how long individual components took to render. Both  1. including any child components and 2. excluding child components.

However, the problem there is how and where to show the results. I'd prolly make most sense using the django-debug-toolbar, e.g. forking / taking inspo from the [template-timings](https://github.com/orf/django-debug-toolbar-template-timings).

**Comment by gone** [comment]
I just got bit with a performance hit, loading a template tag in a component that was rendered multiple times.

**Comment by JuroOravec** [comment]
@gone Could you provide more details? My guess is that relates to the parsing of our `{% component %}` tags, which is currently slow. I've already wrote a Rust port for it, but it still needs to be integrated.

On a different note, if you guys would like to see a speedup in resolving these performance issues, be sure to sponsor this project and give this priority! After spending 3 man days on a [pesky bug](https://github.com/django-components/django-components/pull/956), that I encountered when writing benchmark tests for this issue, had to take a break and work on fun stuff ([syntax highlighting in docs](https://github.com/django-components/pygments-djc), and [Storybook integration](https://github.com/django-components/django-components/issues/967)). But to finish off the storybook integration, I'll need to finally introduce the plugin system, so 3rd party packages can safely intercept when a Component class is created or un/registered. So I reckon I'll get back to these performance-related issues in a week.

**Comment by JuroOravec** [comment]
## Lil update

So I'm still far from being done, but:

- **Current state - Early results**
  - Setup time - roughly the same between django-components and vanilla Django
  - First render - 5x slower than vanilla Django
     - Tho this should improve when https://github.com/django-components/django-components/issues/935 is implemented, and when the Rust-based template tag parser is implemented.
  - Subsequent renders - 4.5x slower than vanilla Django

  Django:
     ```
     Setup:         0.022586 seconds
     First render:  0.104294 seconds
     Second render: 0.047193 seconds
     ```

   Django-components
     ```
     Setup:         0.018866 seconds
     First render:  0.506581 seconds
     Second render: 0.213885 seconds
     ```

    However! I think there's still room for improvement, even without having to solely rely on Rust. For example, I tried to manually do, let's call it "inline optimization", where I inlined constants (parts of the template or component inputs that do not change across). So e.g. if there was a component like icon or button, which had all of its inputs constant, then I replaced the whole component with a plain string.

    With this, I was able to half the time of the second render:
     ```
     Setup:         0.018866 seconds
     First render:  0.506581 seconds
     Second render: 0.105081 seconds
     ```

    So in this case the render time is only 2.2x slower than vanilla Django.

    However, this optimization will become feasible only once in v2 (django-components as a standalone templating engine), because it will require us to own the `{% if %}` and `{% for %}` template tags, among other.

    I outlined the [idea here](https://github.com/django-components/django-components/issues/14#issuecomment-2631211650):
    
    > I also think that there will be optimization opportunities if we take over parsing of whole templates.
    > 
    > * In such case we could for example pre-compile parts of the template to avoid needless resolution.  So if there's a template like this:
    >    ```django
    >   {% for item in menu_list %}
    >     {{ item }}
    >   {% endfor %}
    >   {% component "test" / %}
    >   ```
    >     
    >   And Component looks like this:
    >   ```py
    >   menu_list = ["a", "b", "c", ...]
    >   
    >   class Table(Component):
    >       def get_context_data(self):
    >           return {
    >               "menu_list": menu_list,
    >           }          
    >   ```
    >
    >   Then with a bit of Python-fu, it WOULD be possible to detect that `menu_list` is constant (does not change with component inputs). And with this info, we could modify the Django template, inlining the static data:
    >
    >   ```django
    >     a
    >     b
    >     c
    >   {% component "test" / %}
    >   ```
    >
    >   Thus avoiding having to render the `{% for %}` loop again and again.


- **Taming `asv`** - Turns out I can use `asv` for BOTH validating PRs on whether the performance changed, AND for generating a static site with an overview of all benchmarks. So I'm looking into that over `pytest-benchmark`. The `asv` PR testing will post a comment that may look like this:

   <img width="638" alt="Image" src="https://github.com/user-attachments/assets/78168410-2d5e-42e5-ad6c-4bdca725f6fe" />

**Comment by JuroOravec** [comment]
@oliverhaas Sounds good to me, esp. the approach of reusing Django's API for caching, and basically just allowing to pass the caching settings to Django, so it's as if we wrapped the components in `{% cache %}`.

Let's hear also what @EmilStenstrom thinks. But let's move the discussion on this component-level caching to #992

**Comment by EmilStenstrom** [comment]
@oliverhaas @JuroOravec So why wouldn't this be the default always enabled mode? Why would you have to enable it explicitly per component?

**Comment by jonathan-s** [comment]
> So why wouldn't this be the default always enabled mode?
I think one of the reasons is that caching can have non-intuitive side effects. I think there is a reason for this quote. 

> There are only two hard things in Computer Science: cache invalidation and naming things.

Anyway, I've been following this thread with great interest. It's come a long way since I asked the initial question.

**Comment by EmilStenstrom** [comment]
That's the general answer, yes, but in this case, does it hold? It seems every input to the component will be used to build the cache key. So why would you even need to invalidate it?

**Comment by oliverhaas** [comment]
> [@oliverhaas](https://github.com/oliverhaas) [@JuroOravec](https://github.com/JuroOravec) So why wouldn't this be the default always enabled mode? Why would you have to enable it explicitly per component?

Because of something like this:

```python
class UserIcon(Component):
    cache = True
    def get_context_data(self, user_id):
        user = User.objects.get(id=user_id)
        return {
            "icon_url": user.icon.url,
        } 
```

Because we only hash by the given input parameters (`user_id`), the component would think the template never changes, but we can see that the `user.icon` could be modified somewhere else. Basically all "dumb" components without logic could be cached, as long as input arguments are actually hashable.

But this is only works for `get_context_data`, and would ignore slots as a start. Slots with "isolated" context are probably quite a bit more difficult to get right. Non-isolated slight might be impossible, at least in a useful manner.

**Comment by EmilStenstrom** [comment]
@oliverhaas Mskes sense, great example! 👍 from me too!

**Comment by JuroOravec** [comment]
Closing this one, I won't be pursuing more perf optimisations in django-components (unless critical). New optimisations land in [Citry](https://citry.dev/) ([citry-dev/citry](https://github.com/citry-dev/citry)), the successor to django-components.

See [Citry's benchmarks](https://citry.dev/about/benchmarks/) for how it compares against django-components and Django.
